upb/decode_fast.c

```diff
@@ -360,16 +312,22 @@ static const char *fastdecode_submsg(UPB_PARSE_PARAMS, int tagbytes,
     RETURN_GENERIC("submessage field tag mismatch\n");
   }
 
-  if (--d->depth < 0) return fastdecode_err(d);
+  if (--d->depth == 0) return fastdecode_err(d);
 
   upb_msg **submsg;
   upb_array *arr;
   void *end;
-  uint16_t submsg_idx = data >> 32;
+  uint32_t submsg_idx = data;
+  submsg_idx >>= 16; asm("" : "+r" (submsg_idx));
```

Comment:
What is the purpose of the asm()? Are you fighting the compiler here?

Reply:
Yes fighting the asm. submsg_idx is used in lookup in 8 byte array, instead of
mov eax, data
shr eax, 16
mov ptr, [base + rax * 8]

it folds *8 with the shift
mov eax, data
shr eax, 13
and eax, FFFF8
mov, ptr[base + rax]

Reply:
I removed it